---
title: "R_assignment"
auther: "Nic Lizer"
date: 3/11/22
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Firstly, load up these packages
```{r}
library(dplyr)
library(tidyverse)
library(ggplot2)
library(knitr)
```

##Download the 2 data sets
```{r}
geno<-read_tsv("https://raw.githubusercontent.com/niclizer/BCB_546_UNIX_Assignment/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
snp <- read_tsv("https://raw.githubusercontent.com/niclizer/BCB_546_UNIX_Assignment/main/assignments/UNIX_Assignment/snp_position.txt")
```

# Select SNP_ID, chromosome and Position from SNP_position data set
```{r}
snp_position.selected <- select(snp, c("SNP_ID", "Chromosome", "Position"))
str(snp_position.selected) # 983 rows and 3 columns
#glimpse(snp_position.selected) # similar function with str
```

## Next, determine the number of rows and get infomation of the fang and snp data set
```{r}
nrow(geno)
#2782
ncol(geno)
#986
NROW(na.omit(geno)) 
# 2782
NCOL(na.omit(geno)) 
# 986

object.size(geno)
# 23124352 bytes
object.size(snp)
# 359152 bytes

summary(as.factor(geno$Group))
#TRIPS ZDIPL ZLUXR ZMHUE ZMMIL ZMMLR ZMMMR ZMPBA ZMPIL ZMPJA ZMXCH 
#22    15    17    10   290  1256    27   900    41    34    75 
#ZMXCP ZMXIL ZMXNO ZMXNT ZPERR 
#69     6     7     4     9 
summary(as.factor(snp$Chromosome))
#1=155        10=53        2=127        3=107        4=91        5=122        6=76 #7=97        8=62        9=60 multiple=6  unknown=27 
```
## Then filter by genotype group
```{r}
maize_genotypes <- filter(geno, Group == 'ZMMIL' | Group == 'ZMMLR' | Group == 'ZMMMR')
#dim(maize_genotypes)
#1573  986
#summary(as.factor(maize_genotypes$Group))
#ZMMIL ZMMLR ZMMMR 
#290  1256    27
teosinte_genotypes <- filter(geno, Group == 'ZMPBA' | Group == 'ZMPIL' | Group == 'ZMPJA')
#dim(teosinte_genotypes)
# 975 986
#summary(as.factor(teosinte_genotypes$Group))
#ZMPBA ZMPIL ZMPJA 
#900    41    34
```

## Transpose the both groups sets
```{r}
maize_genotypes <- column_to_rownames(maize_genotypes, var = "Sample_ID")
maize_genotypes.tr <- t(maize_genotypes)%>%as.data.frame()%>%rownames_to_column(., var = "SNP_ID") #transposition step
maize_genotypes.tr <- maize_genotypes.tr[3:nrow(maize_genotypes.tr),] # remove the first 3 rows

### same with teosinte group
teosinte_genotypes <- column_to_rownames(teosinte_genotypes, var = "Sample_ID")
teosinte_genotypes.tr <- t(teosinte_genotypes)%>%as.data.frame()%>%rownames_to_column(., var = "SNP_ID")
teosinte_genotypes.tr <- teosinte_genotypes.tr[3:nrow(teosinte_genotypes.tr),]
```
## Then merge the 2 groups with snp data
```{r}
snp_maizegeno_merge <- merge(snp_position.selected, maize_genotypes.tr, by="SNP_ID")
snp_teosintegeno_merge <- merge(snp_position.selected, teosinte_genotypes.tr, by="SNP_ID")
```
## Then take out the individual chromosomes as well as arrange them in ascending order 
```{r}
maize_chrom_inc1 <- subset(snp_maizegeno_merge, Chromosome==1)%>%arrange(as.numeric(Position))
maize_chrom_inc2 <- subset(snp_maizegeno_merge, Chromosome==2)%>%arrange(as.numeric(Position))
maize_chrom_inc3 <- subset(snp_maizegeno_merge, Chromosome==3)%>%arrange(as.numeric(Position))
maize_chrom_inc4 <- subset(snp_maizegeno_merge, Chromosome==4)%>%arrange(as.numeric(Position))
maize_chrom_inc5 <- subset(snp_maizegeno_merge, Chromosome==5)%>%arrange(as.numeric(Position))
maize_chrom_inc6 <- subset(snp_maizegeno_merge, Chromosome==6)%>%arrange(as.numeric(Position))
maize_chrom_inc7 <- subset(snp_maizegeno_merge, Chromosome==7)%>%arrange(as.numeric(Position))
maize_chrom_inc8 <- subset(snp_maizegeno_merge, Chromosome==8)%>%arrange(as.numeric(Position))
maize_chrom_inc9 <- subset(snp_maizegeno_merge, Chromosome==9)%>%arrange(as.numeric(Position))
maize_chrom_inc10 <- subset(snp_maizegeno_merge, Chromosome==10)%>%arrange(as.numeric(Position))

teosinte_chrom_inc1 <- subset(snp_teosintegeno_merge, Chromosome==1)%>%arrange(as.numeric(Position))
teosinte_chrom_inc2 <- subset(snp_teosintegeno_merge, Chromosome==2)%>%arrange(as.numeric(Position))
teosinte_chrom_inc3 <- subset(snp_teosintegeno_merge, Chromosome==3)%>%arrange(as.numeric(Position))
teosinte_chrom_inc4 <- subset(snp_teosintegeno_merge, Chromosome==4)%>%arrange(as.numeric(Position))
teosinte_chrom_inc5 <- subset(snp_teosintegeno_merge, Chromosome==5)%>%arrange(as.numeric(Position))
teosinte_chrom_inc6 <- subset(snp_teosintegeno_merge, Chromosome==6)%>%arrange(as.numeric(Position))
teosinte_chrom_inc7 <- subset(snp_teosintegeno_merge, Chromosome==7)%>%arrange(as.numeric(Position))
teosinte_chrom_inc8 <- subset(snp_teosintegeno_merge, Chromosome==8)%>%arrange(as.numeric(Position))
teosinte_chrom_inc9 <- subset(snp_teosintegeno_merge, Chromosome==9)%>%arrange(as.numeric(Position))
teosinte_chrom_inc10 <- subset(snp_teosintegeno_merge, Chromosome==10)%>%arrange(as.numeric(Position))
```
###Finally, create the decending files with the replacment of "?" to "-"
```{r}
maize_replaced<- data.frame(lapply(snp_maizegeno_merge, gsub, pattern = "[?]", replacement = "-"))
teosente_replaced <- data.frame(lapply(snp_teosintegeno_merge, gsub, pattern = "[?]", replacement = "-"))

maize_chrom_dec1 <- subset(maize_replaced, Chromosome==1)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec2 <- subset(maize_replaced, Chromosome==2)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec3 <- subset(maize_replaced, Chromosome==3)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec4 <- subset(maize_replaced, Chromosome==4)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec5 <- subset(maize_replaced, Chromosome==5)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec6 <- subset(maize_replaced, Chromosome==6)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec7 <- subset(maize_replaced, Chromosome==7)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec8 <- subset(maize_replaced, Chromosome==8)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec9 <- subset(maize_replaced, Chromosome==9)%>%arrange(desc(as.numeric(Position)))
maize_chrom_dec10 <- subset(maize_replaced, Chromosome==10)%>%arrange(desc(as.numeric(Position)))

teosinte_chrom_dec1 <- subset(teosente_replaced, Chromosome==1)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec2 <- subset(teosente_replaced, Chromosome==2)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec3 <- subset(teosente_replaced, Chromosome==3)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec4 <- subset(teosente_replaced, Chromosome==4)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec5 <- subset(teosente_replaced, Chromosome==5)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec6 <- subset(teosente_replaced, Chromosome==6)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec7 <- subset(teosente_replaced, Chromosome==7)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec8 <- subset(teosente_replaced, Chromosome==8)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec9 <- subset(teosente_replaced, Chromosome==9)%>%arrange(desc(as.numeric(Position)))
teosinte_chrom_dec10 <- subset(teosente_replaced, Chromosome==10)%>%arrange(desc(as.numeric(Position)))
```
### If you would like to actually have a file of these make then you can make csv files with 
```{r}
#write.csv(maize_chrom_inc1,"./maize_chrom_inc1")
#write.csv(maize_chrom_inc2,"./maize_chrom_inc2")
#write.csv(maize_chrom_inc3,"./maize_chrom_inc3")
#write.csv(maize_chrom_inc4,"./maize_chrom_inc4")
#write.csv(maize_chrom_inc5,"./maize_chrom_inc5")
#write.csv(maize_chrom_inc6,"./maize_chrom_inc6")
#write.csv(maize_chrom_inc7,"./maize_chrom_inc7")
#write.csv(maize_chrom_inc8,"./maize_chrom_inc8")
#write.csv(maize_chrom_inc9,"./maize_chrom_inc9")
#write.csv(maize_chrom_inc10,"./maize_chrom_inc10")
#write.csv(teosinte_chrom_inc1,"./teosinte_chrom_inc1")
#write.csv(teosinte_chrom_inc2,"./teosinte_chrom_inc2")
#write.csv(teosinte_chrom_inc3,"./teosinte_chrom_inc3")
#write.csv(teosinte_chrom_inc4,"./teosinte_chrom_inc4")
#write.csv(teosinte_chrom_inc5,"./teosinte_chrom_inc5")
#write.csv(teosinte_chrom_inc6,"./teosinte_chrom_inc6")
#write.csv(teosinte_chrom_inc7,"./teosinte_chrom_inc7")
#write.csv(teosinte_chrom_inc8,"./teosinte_chrom_inc8")
#write.csv(teosinte_chrom_inc9,"./teosinte_chrom_inc9")
#write.csv(teosinte_chrom_inc10,"./teosinte_chrom_inc10")
#write.csv(maize_chrom_dec1,"./maize_chrom_dec1")
#write.csv(maize_chrom_dec2,"./maize_chrom_dec2")
#write.csv(maize_chrom_dec3,"./maize_chrom_dec3")
#write.csv(maize_chrom_dec4,"./maize_chrom_dec4")
#write.csv(maize_chrom_dec5,"./maize_chrom_dec5")
#write.csv(maize_chrom_dec6,"./maize_chrom_dec6")
#write.csv(maize_chrom_dec7,"./maize_chrom_dec7")
#write.csv(maize_chrom_dec8,"./maize_chrom_dec8")
#write.csv(maize_chrom_dec9,"./maize_chrom_dec9")
#write.csv(maize_chrom_dec10,"./maize_chrom_dec10")
#write.csv(teosinte_chrom_dec1,"./teosinte_chrom_dec1")
#write.csv(teosinte_chrom_dec2,"./teosinte_chrom_dec2")
#write.csv(teosinte_chrom_dec3,"./teosinte_chrom_dec3")
#write.csv(teosinte_chrom_dec4,"./teosinte_chrom_dec4")
#write.csv(teosinte_chrom_dec5,"./teosinte_chrom_dec5")
#write.csv(teosinte_chrom_dec6,"./teosinte_chrom_dec6")
#write.csv(teosinte_chrom_dec7,"./teosinte_chrom_dec7")
#write.csv(teosinte_chrom_dec8,"./teosinte_chrom_dec8")
#write.csv(teosinte_chrom_dec9,"./teosinte_chrom_dec9")
#write.csv(teosinte_chrom_dec10,"./teosinte_chrom_dec10")
```

## Part 2 visualization
### To visualize the distrobution of SNPs across chromosomes
```{r}
maize_id <- snp_maizegeno_merge%>% select(SNP_ID, Chromosome, Position) %>% mutate(Species = "Maize")
teosinte_id <- snp_teosintegeno_merge%>% select(SNP_ID, Chromosome, Position) %>% mutate(Species = "Teosinte")
mt_together <- bind_rows(maize_id, teosinte_id) #this combines the two groups
SNPs_per_chrom3 <- ggplot(mt_together) + geom_bar(aes(x= Chromosome, fill = Species)) + ggtitle("Maize and Teosinte SNP's across Chromosomes")
print(SNPs_per_chrom3)
```
### Next we want to visualize the proportion heterozygosity in Groups and in Samples
```{r}
fang_pivot_longer <- geno %>% pivot_longer(!c(Sample_ID, JG_OTU, Group), names_to="SNP_ID", values_to= "NT")
snp_fang <- merge(fang_pivot_longer, snp, by="SNP_ID")
num_snp_fang <- (snp_fang[!is.na(as.numeric(snp_fang$Chromosome)),])


added_column <- num_snp_fang
added_column$Heterozygotes <- "Heterozygotes"
added_column$Heterozygotes[added_column$NT == "?/?"] <- "Missing"
added_column$Heterozygotes[added_column$NT %in% c("A/A", "T/T", "C/C", "G/G")] <- "Homozygotes"

Heterozygosity_in_all_Samples <- (ggplot(added_column, aes(x = Sample_ID, fill = Heterozygotes)) +
  geom_bar(position = "fill") +
  labs(x = "Sample_ID", y = "Proportion") +
  ggtitle("Heterozygosity_in_all_Samples"))
print(Heterozygosity_in_all_Samples)

Heterozygosity_in_all_Groups <- (ggplot(added_column, aes(x = Group, fill = Heterozygotes)) + geom_bar(position = "fill") + labs(x = "Group", y = "Proportion") + theme(axis.text.x = element_text(angle = 90)) +
ggtitle("Heterozygosity_in_all_Groups"))
print(Heterozygosity_in_all_Groups)

```

### My addition is simple, just a graph with only the amount missing in each group.
```{r}

Only_missing <- filter(added_column, Heterozygotes == 'Missing')

only_graph <- ggplot(Only_missing) + geom_bar(aes(x= Group, fill = Heterozygotes))+ labs(x = "Group", y = "Count") + theme(axis.text.x = element_text(angle = 90)) +
ggtitle("Amount_Missing_Only")
print(only_graph)

BBH <- filter(added_column, Heterozygotes == 'Heterozygotes')

```
